### PR TITLE
Add testcloud provisioner

### PR DESCRIPTION
**Pull request by thrix**
This MR adds provisioning of qcow2 images via testcloud.

To test with a qcow2 from url:

    tmt run -a provision -h testcloud --image https://kojipkgs.fedoraproject.org/compose/branched/latest-Fedora-32/compose/Cloud/x86_64/images/Fedora-Cloud-Base-32-20200301.n.0.x86_64.qcow2

To test with a local qcow2:

    tmt run -a provision -h testcloud --image file:///home/user/fedora.qcow2

To test with latest Rawhide:

    tmt run -a provision -h testcloud --image fedora
    tmt run -a provision -h testcloud --image rawhide
    tmt run -a provision -h testcloud --image fedora-rawhide

Other mappings will come later, as it is a bit PITA to do ....

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

### CODE REVIEW

**Comment by thrix** [comment]
Note that testcloud will print on stdout information about formatting the VM disk, here is a patch for that for testcloud:
https://pagure.io/testcloud/pull-request/61

File: tmt.spec [hunk 1]
```diff
@@ -63,6 +64,15 @@ Requires: ansible podman
...
 
+%package testcloud
+Summary: Libvirt (via testcloud) provisioner for the Test Management Tool
+Requires: tmt == %{version}-%{release}
+Requires: ansible testcloud
+
+%description testcloud
+All dependencies of the Test Management Tool required to run tests
+in a libvirt environment provisioned using testcloud.
+
```
Comment (psss) [inline tmt.spec]: Include at least `LICENSE`, otherwise the rpm package is not created.

File: setup.py [hunk 2]
```diff
@@ -33,6 +33,8 @@
 install_requires = [
     'fmf>=0.9.2',
     'click',
+    'requests',
+    'testcloud'
```
Comment (psss) [inline setup.py]: I think these should belong to `extra_requires`.

File: tmt.spec (hunk 1 above)
Comment (thrix) [reply tmt.spec]: ack

Comment (thrix) [reply tmt.spec]: fixing ...

File: setup.py (hunk 2 above)
Comment (thrix) [reply setup.py]: fixing

File: tmt/steps/provision/testcloud.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            self.instance_name, image=self.image)
+
+        # generate ssh key
+        self.run(f'ssh-keygen -f {self.ssh_key} -N ""')
+
+        with open(self.ssh_pubkey, 'r') as pubkey:
+            config.USER_DATA = CLOUDINIT_USER_DATA.format(
+                user_name=self.user, public_key=pubkey.read())
+
+        config.DOWNLOAD_PROGRESS = False
```
Comment (psss) [inline tmt/steps/provision/testcloud.py]: Duplicates line 123?

File: setup.py [hunk 4]
```diff
@@ -33,10 +33,12 @@
 install_requires = [
     'fmf>=0.9.2',
     'click',
+    'requests'
 ]
 extras_require = {
     'docs': ['sphinx', 'sphinx_rtd_theme'],
     'tests': ['pytest', 'python-coveralls', 'mock'],
+    'provision': ['testcloud']
```
Comment (psss) [inline setup.py]: pip fails to install `testcloud` unless `libvirt-devel` is present on the system. How shall we handle this? Add to documentation?

File: tmt/steps/provision/testcloud.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+        # Initialize testcloud image 
+        self.image = testcloud.image.Image(image_url)
+
+        # Show which image we are using
+        self.info('image', f'{self.image.name}', 'green')
+ 
+        status = f'{self.image.name}'
+        if not os.path.exists(self.image.local_path):
+            self.info('status', 'downloading', 'green')
```
Comment (psss) [inline tmt/steps/provision/testcloud.py]: Download failed when running on my laptop:

    Traceback (most recent call last):
    File "/home/psss/.virtualenvs/tmt/lib/python3.7/site-packages/testcloud/image.py", line 126, in _download_remote_image
        with open(local_path + ".part", 'wb') as f:
    PermissionError: [Errno 13] Permission denied: '/var/lib/testcloud/backingstores/Fedora-Cloud-Base-Rawhide-20200228.n.0.x86_64.qcow2.part'

Shall we use a different location for storing the images?

File: setup.py (hunk 4 above)
Comment (thrix) [reply setup.py]: Yes, I will add it to documentation

File: tmt/steps/provision/testcloud.py (hunk 5 above)
Comment (thrix) [reply tmt/steps/provision/testcloud.py]: no, newgrp testcloud or login again .....

Comment (thrix) [reply tmt/steps/provision/testcloud.py]: should we document again?

Comment (thrix) [reply tmt/steps/provision/testcloud.py]: ah right, but if you just install testcloud from pip, I guess you need to create the location by yourself. meh. maybe we should jsut require users to install testcloud and make sure they login again (or use `newgrp testcloud`) ?

Comment (thrix) [reply tmt/steps/provision/testcloud.py]: I want to use these shared locations, as they provide caching of downloaded images ...

Comment (psss) [reply tmt/steps/provision/testcloud.py]: But by default user is not added to the `testcloud` group. Then we should document this as well.

Comment (thrix) [reply tmt/steps/provision/testcloud.py]: hmm, I thought it is ... will document it

Comment (thrix) [reply tmt/steps/provision/testcloud.py]: fixed, testcloud user now shold not be needed. Unfortunately I had to add the domain XML template to the code, as it is added by testcloud package. Seems now we really need to install only the python package ...

**Comment by psss** [comment]
Seems the `StrictHostKeyChecking` should be added to `rsync` commands as well:

    Warning: remote host identification has changed!
    Failed to run command 'rsync -Rvaze "ssh -i ...

**Comment by psss** [comment]
Or shall we include `-o UserKnownHostsFile=/dev/null` in the ssh options as well?

**Comment by thrix** [comment]
@psss all should be fixed, waiting for ssh to be available was added ... options needed for seamless connection via ssh added to everyting ...

Kindly asking for next round of review

**Comment by psss** [comment]
Nice, thanks! The ssh retry fixed the connection problem. One last thing: Perhaps we should remove `-v` from rsync. I don't think it's necessary to print all synced files on the output.

**Comment by thrix** [comment]
> Nice, thanks! The ssh retry fixed the connection problem. One last thing: Perhaps we should remove `-v` from rsync. I don't think it's necessary to print all synced files on the output.

ok

**Comment by thrix** [comment]
@psss done

File: tmt/steps/provision/testcloud.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+            self.image.prepare()
+        except FileNotFoundError:
+            raise GeneralError(
+                f"Could not find image '{self.image.local_path}'")
+
+        self.instance = testcloud.instance.Instance(
+            self.instance_name, image=self.image)
+
+        # generate ssh key
+        self.run(f'ssh-keygen -f {self.ssh_key} -N ""')
```
Comment (psss) [inline tmt/steps/provision/testcloud.py]: Shall we add `openssh` to the `tmt-testcloud` requires?

**Pull request merged**